src/Collection.php

```diff
@@ -315,44 +316,69 @@ public function slice( int $offset, ?int $length = null ): self {
 	/**
 	 * Returns a new collection of differences between another collection or array.
 	 *
+	 * Optional comparison function can be passed to use array_udiff over array_diff if either data contains
+	 * objects. If no $comparator function passed, will match objects by instance (not values.)
+	 *
 	 * @param array<int|string, mixed>|Collection $data
+	 * @param callable|null $comparator The Comparison function to use.
 	 * @return self
 	 * @throws TypeError
 	 */
-	public function diff( $data ):self {
+	public function diff( $data, ?callable $comparator = null ):self {
 
 		if ( ! is_array( $data ) && ! is_a( $data, static::class ) ) {
-			throw new \TypeError( 'Can only merge with other Collections or Arrays.' );
+			throw new \TypeError( 'Can only find the diff with other Collections or Arrays.' );
 		}
 
-		return new static(
-			array_diff(
-				$this->data,
-				is_object( $data ) && is_a( $data, static::class ) ? $data->to_array() : $data
-			)
-		);
+		$data = is_object( $data ) && is_a( $data, static::class ) ? $data->to_array() : $data;
+
+		$new_data = $this->contains_array_or_object( (array) $this->data ) || $this->contains_array_or_object( (array) $data )
```

Comment:
how come you're casting to array if it's already an array above?

Reply:
actually nvm, $this->data vs $data can get confusing if they're different things, I'd rename one as collectionData or the other as comparissonData tbh

Reply:
actually nvm no2 you also cast $data as array, but you shouldn't have to, right?